type annotations required with `-Znew-solver` strsim:0.10.0

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```sh
cargo +nightly new test_strsim
cd test_strsim
cargo +nightly add strsim@0.10.0
RUSTFLAGS="-Znext-solver" cargo +nightly check
```

I expected to see this happen: Code compiles (as without the `-Znew-solver` flag)

Instead, this happened: Compilation failed

```
error[E0282]: type annotations needed
  --> /home/weiznich/.cargo/registry/src/index.crates.io-6f17d22bba15001f/strsim-0.10.0/src/lib.rs:93:13
   |
93 |     for (i, a_elem) in a.into_iter().enumerate() {
   |             ^^^^^^ cannot infer type for associated type `<<&'a Iter1 as IntoIterator>::IntoIter as Iterator>::Item`

error[E0282]: type annotations needed
   --> /home/weiznich/.cargo/registry/src/index.crates.io-6f17d22bba15001f/strsim-0.10.0/src/lib.rs:212:13
    |
212 |     for (i, a_elem) in a.into_iter().enumerate() {
    |             ^^^^^^ cannot infer type for associated type `<<&'a Iter1 as IntoIterator>::IntoIter as Iterator>::Item`

For more information about this error, try `rustc --explain E0282`.

```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.77.0-nightly (f688dd684 2024-01-04)
binary: rustc
commit-hash: f688dd684faca5b31b156fac2c6e0ae81fc9bc90
commit-date: 2024-01-04
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

--- a/tests/ui/traits/next-solver/assembly/param-env-alias-bound-conflict.rs
+++ b/tests/ui/traits/next-solver/assembly/param-env-alias-bound-conflict.rs
@@ -1,0 +1,26 @@
+//@ compile-flags: -Znext-solver
+//@ check-pass
+
+// Regression test for #119607.
+
+pub trait IntoFoo {
+    type Item;
+    type IntoIter: Foo<Item = Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter;
+}
+
+pub trait Foo {
+    type Item;
+
+    fn next(self) -> Option<Self::Item>;
+}
+
+pub fn foo<'a, Iter1, Elem1>(a: &'a Iter1)
+where
+    &'a Iter1: IntoFoo<Item = Elem1>,
+{
+    a.into_iter().next();
+}
+
+fn main() {}

--- a/tests/ui/traits/next-solver/env-shadows-impls/param-env-impl-conflict.rs
+++ b/tests/ui/traits/next-solver/env-shadows-impls/param-env-impl-conflict.rs
@@ -1,0 +1,23 @@
+//@ compile-flags: -Znext-solver
+//@ check-pass
+
+// Regression test for #119608.
+
+pub trait Foo {}
+
+pub trait Bar {
+    type Assoc;
+}
+
+impl<T: Foo> Bar for T {
+    type Assoc = T;
+}
+
+pub fn foo<I>(_input: <I as Bar>::Assoc)
+where
+    I: Bar,
+    <I as Bar>::Assoc: Foo,
+{
+}
+
+fn main() {}

